File: ShortestCycle/ShortestCycle.java
/*
https://www.geeksforgeeks.org/problems/shortest-cycle/1
*/

class Solution {
    public int shortCycle(int V, int[][] edges) {
        ArrayList<ArrayList<Integer>>adjacencyList=new ArrayList<>();
        for(int i=0; i<V; ++i){
            adjacencyList.add(new ArrayList<>());
        }
        for(int edge[]: edges){
            int source=edge[0], destination=edge[1];
            adjacencyList.get(source).add(destination);
            adjacencyList.get(destination).add(source);
        }
        int resultantShortestCycle=Integer.MAX_VALUE;
        for(int node=0; node<V; ++node){
            ArrayList<Integer>distances=new ArrayList<>();
            ArrayList<Integer>parent=new ArrayList<>();
            for(int i=0; i<V; ++i){
                distances.add(-1);
                parent.add(-1);
            }
            LinkedList<Integer>llist=new LinkedList<>();
            llist.add(node);
            distances.set(node, 0);
            while(!llist.isEmpty()){
                int currentLLSize=llist.size();
                while(currentLLSize-- >0){
                    int currentNode=llist.pop();
                    for(int nei: adjacencyList.get(currentNode)){
                        if(distances.get(nei)==-1){
                            distances.set(nei, distances.get(currentNode)+1);
                            parent.set(nei, currentNode);
                            llist.add(nei);
                        } else if(parent.get(currentNode)!=nei){
                            int currentDistance=distances.get(currentNode)+distances.get(nei)+1;
                            resultantShortestCycle=Math.min(resultantShortestCycle, currentDistance);
                        }
                    }
                }
            }
        }
        return resultantShortestCycle==Integer.MAX_VALUE?-1:resultantShortestCycle;
    }
}

File: ShortestCycle/ShortestCycle.cpp
/*
https://www.geeksforgeeks.org/problems/shortest-cycle/1
*/

class Solution {
  public:
    int shortCycle(int V, vector<vector<int>>& edges) {
        vector<vector<int>>adj(V);
        for(auto &edge: edges){
            int src=edge[0], dest=edge[1];
            adj[src].push_back(dest);
            adj[dest].push_back(src);
        }
        int res=INT_MAX;
        for(int node=0; node<V; ++node){
            vector<int>dist(V, -1);
            vector<int>parent(V, -1);
            queue<int>q;
            q.push(node);
            dist[node]=0;
            while(!q.empty()){
                int sz=q.size();
                while(sz--){
                    auto curr=q.front();
                    q.pop();
                    for(auto &nei: adj[curr]){
                        if(dist[nei]==-1){
                            dist[nei]=dist[curr]+1;
                            parent[nei]=curr;
                            q.push(nei);
                        } else if(parent[curr]!=nei){
                            int curDist=dist[curr]+dist[nei]+1;
                            res=min(res, curDist);
                        }
                    }
                }
            }
        }
        return res==INT_MAX?-1:res;
    }
};